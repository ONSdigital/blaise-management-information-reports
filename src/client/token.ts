import { useState } from "react";

type tokenManager = {
  setToken: (userToken: any) => void
  token: any
}

export function getToken(): string | null {
  return localStorage.getItem("token");
}

export function useToken(): tokenManager {
  const [token, setToken] = useState(getToken());

  const saveToken = (userToken: any) => {
    localStorage.setItem("token", userToken);
    setToken(userToken);
  };

  return {
    setToken: saveToken,
    token
  };
}

export function clearToken(): void {
  localStorage.clear();
}

export function authHeader(): Record<string, string> {
  const token = getToken();
  if (!token) {
    return {};
  }

  return {
    authorization: token
  };
}
