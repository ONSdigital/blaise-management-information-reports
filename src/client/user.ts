import axios from "axios";
import { User } from "blaise-api-node-client";

export async function getUser(username: string): Promise<User> {
  const response = await axios.get(`/api/login/users/${username}`);

  return response.data;
}

export async function validatePassword(username: string, password: string): Promise<boolean> {
  const formData = new FormData();
  formData.append("username", username);
  formData.append("password", password);
  const response = await axios.post("/api/login/users/password/validate", formData);

  return response.data;
}
