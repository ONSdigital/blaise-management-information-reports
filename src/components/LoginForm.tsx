import React, { ReactElement } from "react";
import { ONSPanel, StyledForm } from "blaise-design-system-react-components";
import { Component } from "react";
import { validatePassword, validateUserPermissions } from "../client/user";
import { AuthManager } from "../client/token";


type LoginFormProps = {
  authManager: AuthManager
  setLoggedIn: (loggedIn: boolean) => void
}

type LoginFormState = {
  error: string
}

export default class LoginForm extends Component<LoginFormProps, LoginFormState> {
  constructor(props: LoginFormProps) {
    super(props);
    this.login = this.login.bind(this);
    this.state = {
      error: ""
    };
  }

  formFields(): any {
    return [
      {
        name: "Username",
        id: "username",
        description: "Your Blaise username",
        type: "username",
      },
      {
        name: "Password",
        description: "Your Blaise password",
        type: "password",
      }
    ];
  }

  async login(form: Record<string, string>, setSubmitting: (isSubmitting: boolean) => void): Promise<void> {
    const valid = await validatePassword(form.Username, form.Password);
    if (!valid) {
      this.setState({
        error: "Incorrect username or password"
      });
      setSubmitting(false);
      return;
    }
    const [authorised, token] = await validateUserPermissions(form.Username);
    if (!authorised) {
      this.setState({
        error: "You do not have the correct permissions"
      });
      setSubmitting(false);
      return;
    }
    this.props.authManager.setToken(token);
    this.props.setLoggedIn(authorised);
  }

  error(): ReactElement | undefined {
    if (this.state.error != "") {
      return <ONSPanel status="error">{this.state.error}</ONSPanel>;
    }
    return undefined;
  }

  render(): ReactElement {
    return (
      <>
        <h1 className="u-mt-m">Sign in</h1>
        {this.error()}
        <StyledForm fields={this.formFields()} onSubmitFunction={this.login} submitLabel="Sign in" />
      </>
    );
  }
}
