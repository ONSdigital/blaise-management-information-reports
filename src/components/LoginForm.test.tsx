
import React from "react";
import { createMemoryHistory } from "history";
import { cleanup, render, waitFor } from "@testing-library/react";
import LoginForm from "./LoginForm";
import { Router } from "react-router";
import { screen } from "@testing-library/dom";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import MockAdapter from "axios-mock-adapter";
import axios from "axios";
import flushPromises from "../tests/utilities";
import jwt from "jsonwebtoken";
import { AuthManager } from "../client/token";

const mockAdapter = new MockAdapter(axios);

let loggedIn = false;

function setLoggedIn(isLoggedIn: boolean) {
  loggedIn = isLoggedIn;
}

describe("Login form", () => {
  const authManager = new AuthManager();

  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
    setLoggedIn(false);
    mockAdapter.reset();
  });

  it("matches snapshot", async () => {
    const history = createMemoryHistory();
    const wrapper = render(
      <Router history={history}>
        <LoginForm authManager={authManager} setLoggedIn={setLoggedIn} />
      </Router>
    );

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  it("renders correctly", async () => {
    const history = createMemoryHistory();
    render(
      <Router history={history}>
        <LoginForm authManager={authManager} setLoggedIn={setLoggedIn} />
      </Router>
    );

    await waitFor(() => {
      expect(screen.queryAllByText("Sign in")).toHaveLength(2);
      expect(screen.queryByText("Username")).toBeVisible();
      expect(screen.queryByText("Password")).toBeVisible();
    });
  });

  describe("when the username or password is incorrect", () => {
    it("renders an error and does not set the token", async () => {
      mockAdapter.onPost("/api/login/users/password/validate").reply(200, false);

      const history = createMemoryHistory();
      render(
        <Router history={history}>
          <LoginForm authManager={authManager} setLoggedIn={setLoggedIn} />
        </Router>
      );

      userEvent.type(screen.getByLabelText("Username"), "test");
      userEvent.type(screen.getByLabelText("Password"), "test");

      userEvent.click(screen.getByTestId("submit-button"));

      await waitFor(() => {
        expect(screen.queryByText("Incorrect username or password")).toBeVisible();
      });

      expect(loggedIn).toBeFalsy();
    });
  });


  describe("when the username and password are correct but the user does not have permission", () => {
    it("renders an error and does not set the token", async () => {
      mockAdapter.onPost("/api/login/users/password/validate").reply(200, true);
      mockAdapter.onGet("/api/login/users/test/authorised").reply(403, { "error": "Not authorised" });

      const history = createMemoryHistory();
      render(
        <Router history={history}>
          <LoginForm authManager={authManager} setLoggedIn={setLoggedIn} />
        </Router>
      );

      userEvent.type(screen.getByLabelText("Username"), "test");
      userEvent.type(screen.getByLabelText("Password"), "test");

      userEvent.click(screen.getByTestId("submit-button"));

      await waitFor(() => {
        expect(screen.queryByText("You do not have the correct permissions")).toBeVisible();
      });

      expect(loggedIn).toBeFalsy();
    });
  });

  describe("when the username and password are correct and the user has permission", () => {
    it("sets the token", async () => {
      mockAdapter.onPost("/api/login/users/password/validate").reply(200, true);
      mockAdapter.onGet("/api/login/users/test/authorised").reply(200, { token: jwt.sign({ data: { "role": "test" } }, "test-secret") });

      const history = createMemoryHistory();
      render(
        <Router history={history}>
          <LoginForm authManager={authManager} setLoggedIn={setLoggedIn} />
        </Router>
      );

      userEvent.type(screen.getByLabelText("Username"), "test");
      userEvent.type(screen.getByLabelText("Password"), "test");

      userEvent.click(screen.getByTestId("submit-button"));

      await waitFor(async () => {
        await flushPromises();
      });

      expect(loggedIn).toBeTruthy();
    });
  });
});
