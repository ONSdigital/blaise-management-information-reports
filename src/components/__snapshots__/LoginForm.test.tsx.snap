// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login form matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1
        class="u-mt-m"
      >
        Sign in
      </h1>
      <form
        action="#"
      >
        <div
          class="field"
        >
          <label
            class="label label--with-description"
            for="username"
          >
            Username
          </label>
          <span
            class="label__description  input--with-description"
            id="description-hint"
          >
            Your Blaise username
          </span>
          <input
            class="input input--text input-type__input "
            id="username"
            name="Username"
            type="username"
            value=""
          />
        </div>
        <div
          class="field"
        >
          <label
            class="label label--with-description"
            for="Password"
          >
            Password
          </label>
          <span
            class="label__description  input--with-description"
            id="description-hint"
          >
            Your Blaise password
          </span>
          <input
            class="input input--text input-type__input "
            id="Password"
            name="Password"
            type="password"
            value=""
          />
        </div>
        <br />
        <button
          class="btn "
          data-testid="submit-button"
          type="submit"
        >
          <span
            class="btn__inner"
          >
            Sign in
          </span>
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <h1
      class="u-mt-m"
    >
      Sign in
    </h1>
    <form
      action="#"
    >
      <div
        class="field"
      >
        <label
          class="label label--with-description"
          for="username"
        >
          Username
        </label>
        <span
          class="label__description  input--with-description"
          id="description-hint"
        >
          Your Blaise username
        </span>
        <input
          class="input input--text input-type__input "
          id="username"
          name="Username"
          type="username"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          class="label label--with-description"
          for="Password"
        >
          Password
        </label>
        <span
          class="label__description  input--with-description"
          id="description-hint"
        >
          Your Blaise password
        </span>
        <input
          class="input input--text input-type__input "
          id="Password"
          name="Password"
          type="password"
          value=""
        />
      </div>
      <br />
      <button
        class="btn "
        data-testid="submit-button"
        type="submit"
      >
        <span
          class="btn__inner"
        >
          Sign in
        </span>
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
